### ISSUE TITLE
if.bind renders DOM element and subsequently removes the element after binding

### ISSUE THREAD

**Issue by nCubed** [aurelia/templating#363]
Is the expected behavior of `if.bind`:
1. Render the HTML of the template
2. Evaluate the binding expression
3. Remove the element from the DOM if the binding expression is false

For example, I would like to only present an image if the binding property of the image has a value:

``` html
<div repeat.for="m of items">
    <img if.bind="m.image" src="http://example.com/images/${m.image}" />
    <div>${m.title}</div>
</div>
```

I've also tried to wrap the image within a div container:

``` html
<div repeat.for="m of items">
    <div if.bind="m.image"><img src="http://example.com/images/${m.image}" /></div>
    <div>${m.title}</div>
</div>
```

In both cases, the same behavior occurs where the `<img />` element is being output to the browser resulting in a _400 Bad Request_ of `http://example.com/images/`. And after the `<img />` is rendered, the element is subsequently removed from the DOM when `if.bind` returns false.

**Comment by EisenbergEffect** [comment]
This definitely should not be the behavior. @martingust @jdanyow There was a recent change made in `if`. Is it possible that it caused this issue?

**Comment by nCubed** [comment]
I pulled down the latest getting started es 2106 kit and can not reproduce this issue. I've been following the pluralsight tutorial; perhaps I've set something up incorrectly. 

For now, I'd say this issue should be on-hold. I'm going to take the es 2016 getting started kit and see if I can reproduce the same issue I encountered with the pluralsight tutorial.

**Comment by EisenbergEffect** [comment]
The getting started kit may have some slightly older versions of dependencies from the latest patch releases, so it's possible the issue still exists.

**Comment by nCubed** [comment]
I've pulled down the latest and have been able to reproduce the issue. From what I can tell, the issue only occurs when modifying the binding context. I've attached a working sample for your review.

1. Install the missing packages
1. Run the app
1. Go to the issue363 tab
1. Use the next button to page to the 2nd page of results

Notice the initial page will have 1 missing image for "Lea to the Rescue", but not a 400 bad request. Upon going to the next/previous pages, all missing images will have a 400 bad request.

Perhaps I've set something up incorrectly with the paged results?

[issue-363.zip](https://github.com/aurelia/templating/files/289950/issue-363.zip)

**Comment by EisenbergEffect** [comment]
@jdanyow Can you or @martingust review this?

**Comment by nCubed** [comment]
If this helps, I've re-created a much simpler example:

``` javascript
constructor() {
    this.images = [
        "https://avatars0.githubusercontent.com/u/131485?v=3&s=96"
        , null
        , "https://avatars2.githubusercontent.com/u/1033214?v=3&s=96"
        , null
        , null
    ];

    this.index = 0;
    this.image = this.images[this.index];
}

next() {
    this.index = ++this.index === this.images.length ? 0 : this.index;
    this.image = this.images[this.index];
}
```

```html
<div>
  <button type="button" 
          class="btn btn-success" 
          click.delegate="next()">next</button>
  <img src.bind="image" if.bind="image" />
  <p>images index: ${index} | image value = ${image}</p>
</div>
```

**Comment by jdanyow** [comment]
https://gist.run/?id=69effd1d55176a78d2767ed65bfdfd99

**Comment by nCubed** [comment]
@jdanyow - using the gist you just created, open up the developer tools, view the console, and you'll see when you click through to the next image(s), a 404 error is returned when the previous image had a value and the new image is null.

**Comment by jdanyow** [comment]
yep- was just adding the gist for reference.

I've debugged through this, and tested with the previous version of the `if` behavior.  Turns out this is not a new bug, it's always behaved like this.

What happens when the view model's `image` property is assigned null, [this queues a task to notify the subscribers](https://github.com/aurelia/binding/blob/master/src/property-observation.js#L57-L69).

There are three subscribers which are called in this order:
1. an interpolation binding `... image value = ${image}</p>` which we don't care about
2. the if binding's value which is a `BehaviorPropertyObserver`... [this code gets called](https://github.com/aurelia/templating/blob/master/src/behavior-property-observer.js#L38-L50) which queues a micro task (meaning the `if`'s `valueChanged` logic is now at the "end of the line")
3. the image.src binding

tldr: by the time the `if` logic executes and unbinds it's template, the image's src prop has been assigned, causing the image load attempt/404.

@EisenbergEffect thoughts?

**Comment by EisenbergEffect** [comment]
Off hand, I'm not sure of a good way to fix this. Basically, we would have to have a way to get the if binding to update first, correct? Perhaps we can prioritize template controller attributes above other attributes by ordering them first in the bindings array so that they get enqueued before normal bindings?

**Comment by EisenbergEffect** [comment]
Although, that doesn't really make sense not that I think about it because the attributes are in two different views...

**Comment by EisenbergEffect** [comment]
@jdanyow I guess my question is...why is the `if` attribute creating the view at all to begin with?

**Comment by jdanyow** [comment]
@EisenbergEffect because the `image` property is initially truthy the `if` renders it's template.  Later, the `image` property is assigned null.  Because of the indirection with the task queue and behavior property observer, the image element's src binding updates before the `if` controller finds out it needs to detach and unbind it's view.

**Comment by EisenbergEffect** [comment]
Oops. I wasn't looking at the src that was showing the swapping of the values. That makes sense.

I'm not sure if there's an easy way to fix this. Do you recall the reason for using the task queue? Is there some way we can know when we have to use that vs. not, so that we can avoid the timing issue? It's definitely odd that the if evaluates after its own template. We really don't want that to happen.

**Comment by EisenbergEffect** [comment]
Perhaps @martingust has some thoughts on this.

**Comment by jdanyow** [comment]
it's the task queue usage in the BehaviorPropertyObserver

**Comment by leiavoia** [comment]
I've run into this issue several times just getting started with aurelia. It seems most intuitive for an if.bind to run first before anything else. Besides the image 404 example originally presented, i also hit a similar issue with dynamic composition. 

This is a dynamic composition test. The objective is to switch content out based on a template name. However, when the template name == false, it fails with error NetworkError: 404 Not Found - http://localhost:9000/src/false.js and Unhandled rejection Error: Script error for "false". With newbie logic, the if.bind statement on the compose element should have prevented this. Minimal test case: 

```
<!-- ./widget.js -->
export class Widget { }

<!-- ./widget.html -->
<template>
	<slot></slot>
</template>

<!-- ./app.js -->
export class App {
  template = 'widget';
  ToggleWidget() {
    this.template = this.template == 'widget' ? false : 'widget';
  }
}

<!-- ./app.html -->
<template>
  <require from="./widget"></require>
  <compose if.bind="template == 'widget'" view-model="./${template}">[This is a widget. It should only show if app.template == 'widget']</compose>
  <p><a href="" click.delegate="ToggleWidget()">TOGGLE WIDGET</a></p>
</template>
```

**Comment by doubleaa93** [comment]
Has this been resolved? I'm still having issues caused by the template rendering and then getting removed if condition is false. Using latest

**Comment by jdanyow** [comment]
@doubleaa93 yes- this was resolved- this issue should have been closed. The issues reported here are no longer reproducible with the latest bits.

What you've described (the template rendering and then getting removed if condition is false) is the what the `if` template controller does: adds or removes the DOM element depending on the condition's result. Try using `show` instead if you don't want the element removed.
